sphinx_needs/lsp/esbonio.py

```diff
@@ -0,0 +1,378 @@
+                                ),
+                            ),
+                            new_text="",
+                        )
+                    ],
+                )
+                for need in ls.needs_store.needs_per_doc[requested_doc]
+                if need["type"] == requested_type
+            ]
+
+
+def generate_hash(user_name, doc_uri, need_prefix, line_number):
+    salt = os.urandom(blake2b.SALT_SIZE)  # pylint: disable=no-member
+    return blake2b(
+        f"{user_name}{doc_uri}{need_prefix}{line_number}".encode(),
+        digest_size=4,
+        salt=salt,
+    ).hexdigest()
+
+
+def generate_need_id(ls, params, lines: List[str], word: str, need_type: str = None) -> str:
+    """Generate a need ID including hash suffix."""
+
+    user_name = getpass.getuser()
+    doc_uri = params.doc.uri
+    line_number = params.position.line
+
+    if not need_type:
+        try:
+            match = re.search(".. ([a-z]+)::", lines[line_number - 1])
+            need_type = match.group(1)
+        except AttributeError:
+            return "ID"
+
+    need_prefix = need_type.upper()
+    hash_part = generate_hash(user_name, doc_uri, need_prefix, line_number)
+    need_id = need_prefix + "_" + hash_part
+    # re-generate hash if ID is already in use
+    while need_id in ls.needs_store.needs:
+        hash_part = generate_hash(user_name, doc_uri, need_prefix, line_number)
+        need_id = need_prefix + "_" + hash_part
+    return need_id
+
+
+def complete_directive(ls, params, lines: List[str], word: str):
+    # need_type ~ req, work, act, ...
+    items = []
+    for need_type, title in ls.needs_store.declared_types.items():
+        text = (
+            " " + need_type + ":: ${1:title}\n"
+            "\t:id: ${2:" + generate_need_id(ls, params, lines, word, need_type=need_type) + "}\n"
+            "\t:status: open\n\n"
+            "\t${3:content}.\n$0"
+        )
+        label = f".. {need_type}::"
+        items.append(
+            CompletionItem(
+                label=label,
+                detail=title,
+                insert_text=text,
+                insert_text_format=InsertTextFormat.Snippet,
+                kind=CompletionItemKind.Snippet,
+            )
+        )
+    return items
+
+
+def complete_role_or_option(ls, params, lines: List[str], word: str):
+    return [
+        CompletionItem(
+            label=":id:",
+            detail="needs option",
+            insert_text="id: ${1:" + generate_need_id(ls, params, lines, word) + "}\n$0",
+            insert_text_format=InsertTextFormat.Snippet,
+            kind=CompletionItemKind.Snippet,
+        ),
+        CompletionItem(
+            label=":need:",
+            detail="need role",
+            insert_text="need:`${1:ID}` $0",
+            insert_text_format=InsertTextFormat.Snippet,
+            kind=CompletionItemKind.Snippet,
+        ),
+    ]
+
+
+class NeedlsFeatures(LanguageFeature):
```

Comment:
class name should use PascalCase:

```suggestion
class NeedLsFeatures(LanguageFeature):
```